titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultDirectKubeApiServerIntegrator.java

```diff
@@ -140,27 +144,19 @@ public boolean isReadyForScheduling() {
     @Override
     public Mono<V1Pod> launchTask(Job job, Task task) {
         return Mono.fromCallable(() -> {
-            Stopwatch timer = Stopwatch.createStarted();
             try {
                 V1Pod v1Pod = taskToPodConverter.apply(job, task);
                 logger.info("creating pod: {}", v1Pod);
-
-                fitKubeInjection.ifPresent(i -> i.beforeImmediate(KubeFitAction.ErrorKind.POD_CREATE_ERROR.name()));
-
-                kubeApiFacade.getCoreV1Api().createNamespacedPod(KUBERNETES_NAMESPACE, v1Pod, null, null, null);
-                pods.putIfAbsent(task.getId(), v1Pod);
-
-                metrics.launchSuccess(task, v1Pod, timer.elapsed(TimeUnit.MILLISECONDS));
-
                 return v1Pod;
             } catch (Exception e) {
-                logger.error("Unable to create pod with error: {}", KubeUtil.toErrorDetails(e), e);
-
-                metrics.launchError(task, e, timer.elapsed(TimeUnit.MILLISECONDS));
-
-                throw new IllegalStateException("Unable to launch a task " + task.getId(), e);
+                logger.error("Unable to convert job {} and task {} to pod: {}", job, task, KubeUtil.toErrorDetails(e), e);
+                throw new IllegalStateException("Unable to convert task to pod " + task.getId(), e);
             }
-        }).subscribeOn(apiClientScheduler).timeout(Duration.ofMillis(configuration.getKubeApiClientTimeoutMs()));
+        })
+                .flatMap(v1Pod -> launchPersistentVolume(job, task, v1Pod, kubeApiFacade.getCoreV1Api())
+                        .then(Mono.just(v1Pod)))
+                .flatMap(v1Pod -> launchPod(task, v1Pod))
+                .subscribeOn(apiClientScheduler).timeout(Duration.ofMillis(configuration.getKubeApiClientTimeoutMs()));
```

Comment:
Do we need to increase the timeout now that we are adding a call to `launchPersistentVolume`?

Reply:
That's a good question. My impression is that the default 5 seconds will be sufficient for both create operations (i.e., 2 ApiServer create/write calls) based on the average latency we are seeing in today. Though the timeout is a FP, so we can adjust as collect more experience.